File: backend/src/services/analysis.service.ts
import { promises as fs } from 'fs'
import path from 'path'

export interface CodeIssue {
  id: string
  type: 'bug' | 'security' | 'performance' | 'style' | 'complexity'
  severity: 'low' | 'medium' | 'high' | 'critical'
  message: string
  line: number
  column?: number
  endLine?: number
  endColumn?: number
  code?: string
}

export interface CodeMetrics {
  linesOfCode: number
  complexity: number
  maintainabilityIndex: number
  technicalDebt: number
  dependencies: string[]
  duplicateCode: number
}

export class AnalysisService {
  async analyzeFile(filePath: string, language: string): Promise<{
    issues: CodeIssue[]
    metrics: CodeMetrics
  }> {
    const content = await fs.readFile(filePath, 'utf-8')
    const lines = content.split('\n')

    const issues = await this.detectIssues(content, lines, language)
    const metrics = await this.calculateMetrics(content, lines, language)

    return { issues, metrics }
  }

  private async detectIssues(
    content: string,
    lines: string[],
    language: string
  ): Promise<CodeIssue[]> {
    const issues: CodeIssue[] = []

    // Simple pattern-based analysis (in production, use proper AST analysis)
    const patterns = {
      // Security issues
      eval: {
        pattern: /eval\(/g,
        type: 'security' as const,
        severity: 'critical' as const,
        message: 'Use of eval() is a security risk',
      },
      sqlInjection: {
        pattern: /query\([^)]*\+[^)]*\)/g,
        type: 'security' as const,
        severity: 'high' as const,
        message: 'Possible SQL injection vulnerability',
      },
      // Performance issues
      nestedLoops: {
        pattern: /for\s*\([^)]*\)[^{]*{[^}]*for\s*\(/g,
        type: 'performance' as const,
        severity: 'medium' as const,
        message: 'Nested loops may cause performance issues',
      },
      // Complexity issues
      longFunction: {
        threshold: 50,
        type: 'complexity' as const,
        severity: 'medium' as const,
        message: 'Function is too long and should be refactored',
      },
      // Style issues
      console: {
        pattern: /console\.(log|error|warn)/g,
        type: 'style' as const,
        severity: 'low' as const,
        message: 'Console statement should be removed in production',
      },
    }

    // Check each line
    lines.forEach((line, index) => {
      const lineNumber = index + 1

      // Check patterns
      Object.entries(patterns).forEach(([key, rule]) => {
        if ('pattern' in rule) {
          const matches = line.match(rule.pattern)
          if (matches) {
            issues.push({
              id: `${key}-${lineNumber}`,
              type: rule.type,
              severity: rule.severity,
              message: rule.message,
              line: lineNumber,
              code: line.trim(),
            })
          }
        }
      })
    })

    // Check function length
    const functionMatches = content.matchAll(/function\s+\w+\s*\([^)]*\)\s*{/g)
    for (const match of functionMatches) {
      const startIndex = match.index!
      const startLine = content.substring(0, startIndex).split('\n').length

      // Find matching closing brace (simplified)
      const endMatch = content.indexOf('}', startIndex)
      const endLine = content.substring(0, endMatch).split('\n').length

      const functionLength = endLine - startLine
      if (functionLength > 50) {
        issues.push({
          id: `long-func-${startLine}`,
          type: 'complexity',
          severity: 'medium',
          message: `Function is ${functionLength} lines long and should be refactored`,
          line: startLine,
          endLine,
        })
      }
    }

    return issues
  }

  private async calculateMetrics(
    content: string,
    lines: string[],
    language: string
  ): Promise<CodeMetrics> {
    const linesOfCode = lines.filter(line => line.trim() && !line.trim().startsWith('//')).length

    // Calculate cyclomatic complexity (simplified)
    const complexity = this.calculateComplexity(content)

    // Calculate maintainability index (simplified formula)
    const halsteadVolume = Math.log2(linesOfCode + 1) * 10
    const maintainabilityIndex = Math.max(
      0,
      (171 - 5.2 * Math.log(halsteadVolume) - 0.23 * complexity - 16.2 * Math.log(linesOfCode)) * 100 / 171
    )

    // Extract dependencies (simplified)
    const dependencies = this.extractDependencies(content, language)

    // Estimate technical debt in hours
    const technicalDebt = Math.floor(complexity / 10)

    // Check for duplicate code (simplified - would use proper duplicate detection in production)
    const duplicateCode = 0

    return {
      linesOfCode,
      complexity,
      maintainabilityIndex: Math.round(maintainabilityIndex),
      technicalDebt,
      dependencies,
      duplicateCode,
    }
  }

  private calculateComplexity(content: string): number {
    // Simplified cyclomatic complexity calculation
    let complexity = 1 // Base complexity

    const patterns = [
      /if\s*\(/g,
      /else\s+if\s*\(/g,
      /for\s*\(/g,
      /while\s*\(/g,
      /case\s+/g,
      /catch\s*\(/g,
      /\?\s*[^:]+:/g, // Ternary operators
      /&&/g,
      /\|\|/g,
    ]

    patterns.forEach(pattern => {
      const matches = content.match(pattern)
      if (matches) {
        complexity += matches.length
      }
    })

    return complexity
  }

  private extractDependencies(content: string, language: string): string[] {
    const dependencies: Set<string> = new Set()

    if (language === 'javascript' || language === 'typescript') {
      // import statements
      const importMatches = content.matchAll(/import\s+.*?from\s+['"]([^'"]+)['"]/g)
      for (const match of importMatches) {
        dependencies.add(match[1])
      }

      // require statements
      const requireMatches = content.matchAll(/require\s*\(\s*['"]([^'"]+)['"]\s*\)/g)
      for (const match of requireMatches) {
        dependencies.add(match[1])
      }
    } else if (language === 'python') {
      const importMatches = content.matchAll(/^import\s+(\w+)/gm)
      for (const match of importMatches) {
        dependencies.add(match[1])
      }

      const fromMatches = content.matchAll(/^from\s+(\w+)\s+import/gm)
      for (const match of fromMatches) {
        dependencies.add(match[1])
      }
    }

    return Array.from(dependencies)
  }
}

File: backend-flask/services/analysis_service.py
"""Code analysis service for static code analysis"""
import os
import re
import json
from typing import Dict, List, Any
from pathlib import Path


class CodeAnalyzer:
    """Static code analyzer for detecting code issues and metrics"""

    # File extensions to analyze
    SUPPORTED_EXTENSIONS = {
        '.py': 'python',
        '.js': 'javascript',
        '.jsx': 'javascript',
        '.ts': 'typescript',
        '.tsx': 'typescript',
        '.java': 'java',
        '.cpp': 'cpp',
        '.c': 'c',
        '.go': 'go',
        '.rb': 'ruby',
        '.php': 'php',
        '.rs': 'rust',
        '.swift': 'swift',
        '.kt': 'kotlin',
        '.cs': 'csharp'
    }

    # Common code smell patterns
    CODE_SMELLS = {
        'long_function': {
            'pattern': r'def\s+\w+\([^)]*\):|function\s+\w+\([^)]*\)\s*{',
            'threshold': 50,  # lines
            'severity': 'medium',
            'message': 'Function is too long (>50 lines)'
        },
        'deep_nesting': {
            'pattern': r'(\s{4,})(if|for|while|switch)',
            'threshold': 4,  # levels
            'severity': 'medium',
            'message': 'Deep nesting detected (>4 levels)'
        },
        'magic_numbers': {
            'pattern': r'[^a-zA-Z_]\d{2,}[^a-zA-Z_]',
            'severity': 'low',
            'message': 'Magic number detected - consider using named constant'
        },
        'console_log': {
            'pattern': r'console\.(log|debug|info|warn|error)',
            'severity': 'low',
            'message': 'Console statement found - remove before production'
        },
        'todo_comment': {
            'pattern': r'(TODO|FIXME|HACK|XXX):?',
            'severity': 'low',
            'message': 'TODO/FIXME comment found'
        },
        'empty_catch': {
            'pattern': r'catch\s*\([^)]*\)\s*{\s*}',
            'severity': 'high',
            'message': 'Empty catch block - errors are being silently ignored'
        }
    }

    # Security patterns
    SECURITY_ISSUES = {
        'sql_injection': {
            'pattern': r'(execute|query)\s*\(\s*["\'].*\+.*["\']',
            'severity': 'critical',
            'message': 'Potential SQL injection - use parameterized queries'
        },
        'hardcoded_password': {
            'pattern': r'(password|pwd|passwd)\s*=\s*["\'][^"\']+["\']',
            'severity': 'critical',
            'message': 'Hardcoded password detected'
        },
        'eval_usage': {
            'pattern': r'\beval\s*\(',
            'severity': 'critical',
            'message': 'eval() usage detected - potential security risk'
        },
        'weak_crypto': {
            'pattern': r'(md5|sha1)\s*\(',
            'severity': 'high',
            'message': 'Weak cryptographic function - use SHA-256 or better'
        }
    }

    def __init__(self, project_path: str):
        self.project_path = project_path

    def analyze_project(self) -> Dict[str, Any]:
        """Analyze entire project and return results"""
        results = []
        total_issues = 0
        total_files = 0

        for root, dirs, files in os.walk(self.project_path):
            # Skip common directories
            dirs[:] = [d for d in dirs if d not in ['.git', 'node_modules', '__pycache__', 'venv', '.venv', 'dist', 'build']]

            for file in files:
                ext = os.path.splitext(file)[1]
                if ext in self.SUPPORTED_EXTENSIONS:
                    file_path = os.path.join(root, file)
                    relative_path = os.path.relpath(file_path, self.project_path)

                    try:
                        analysis = self.analyze_file(file_path, relative_path)
                        if analysis:
                            results.append(analysis)
                            total_issues += len(analysis['issues'])
                            total_files += 1
                    except Exception as e:
                        print(f'Error analyzing {relative_path}: {e}')

        return {
            'files_analyzed': total_files,
            'total_issues': total_issues,
            'results': results
        }

    def analyze_file(self, file_path: str, relative_path: str) -> Dict[str, Any]:
        """Analyze a single file"""
        try:
            with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                content = f.read()
                lines = content.split('\n')

            ext = os.path.splitext(file_path)[1]
            language = self.SUPPORTED_EXTENSIONS.get(ext, 'unknown')

            issues = []
            suggestions = []

            # Detect code smells
            for smell_name, smell_config in self.CODE_SMELLS.items():
                detected = self._detect_pattern(content, lines, smell_config, smell_name)
                issues.extend(detected)

            # Detect security issues
            for sec_name, sec_config in self.SECURITY_ISSUES.items():
                detected = self._detect_pattern(content, lines, sec_config, sec_name)
                issues.extend(detected)

            # Calculate metrics
            metrics = self._calculate_metrics(content, lines)

            # Generate suggestions based on issues
            if len(issues) > 0:
                suggestions.append({
                    'type': 'refactor',
                    'message': f'Found {len(issues)} issues that could be addressed',
                    'priority': self._calculate_priority(issues)
                })

            return {
                'file_path': relative_path,
                'language': language,
                'issues': issues,
                'suggestions': suggestions,
                'metrics': metrics
            }

        except Exception as e:
            return None

    def _detect_pattern(self, content: str, lines: List[str], config: Dict, issue_type: str) -> List[Dict]:
        """Detect issues based on regex pattern"""
        issues = []
        pattern = config['pattern']

        for i, line in enumerate(lines, start=1):
            matches = re.finditer(pattern, line)
            for match in matches:
                # Special handling for deep nesting
                if issue_type == 'deep_nesting':
                    indent_level = len(match.group(1)) // 4
                    if indent_level >= config.get('threshold', 4):
                        issues.append({
                            'type': issue_type,
                            'line': i,
                            'column': match.start() + 1,
                            'severity': config['severity'],
                            'message': config['message'],
                            'code': line.strip()
                        })
                # Special handling for long functions
                elif issue_type == 'long_function':
                    func_lines = self._count_function_lines(lines, i - 1)
                    if func_lines > config.get('threshold', 50):
                        issues.append({
                            'type': issue_type,
                            'line': i,
                            'column': match.start() + 1,
                            'severity': config['severity'],
                            'message': f'Function is {func_lines} lines long',
                            'code': line.strip()
                        })
                else:
                    issues.append({
                        'type': issue_type,
                        'line': i,
                        'column': match.start() + 1,
                        'severity': config['severity'],
                        'message': config['message'],
                        'code': line.strip()
                    })

        return issues

    def _count_function_lines(self, lines: List[str], start_idx: int) -> int:
        """Count lines in a function starting from start_idx"""
        count = 0
        brace_count = 0
        in_function = False

        for i in range(start_idx, len(lines)):
            line = lines[i].strip()
            if not in_function and ('{' in line or ':' in line):
                in_function = True

            if in_function:
                count += 1
                brace_count += line.count('{') - line.count('}')

                # Python function end detection
                if ':' in lines[start_idx] and i > start_idx:
                    if line and not line.startswith(' ') and not line.startswith('\t'):
                        break

                # Brace-based language function end
                if brace_count == 0 and '{' in lines[start_idx]:
                    break

                if count > 200:  # Safety limit
                    break

        return count

    def _calculate_metrics(self, content: str, lines: List[str]) -> Dict[str, Any]:
        """Calculate code metrics"""
        non_empty_lines = [l for l in lines if l.strip() and not l.strip().startswith(('/', '#', '*'))]

        return {
            'lines_of_code': len(non_empty_lines),
            'total_lines': len(lines),
            'comment_lines': len(lines) - len(non_empty_lines),
            'complexity_estimate': self._estimate_complexity(content),
            'average_line_length': sum(len(l) for l in lines) / max(len(lines), 1)
        }

    def _estimate_complexity(self, content: str) -> int:
        """Rough cyclomatic complexity estimate"""
        complexity = 1  # Base complexity

        # Count decision points
        complexity += len(re.findall(r'\bif\b', content))
        complexity += len(re.findall(r'\belse\b', content))
        complexity += len(re.findall(r'\bfor\b', content))
        complexity += len(re.findall(r'\bwhile\b', content))
        complexity += len(re.findall(r'\bcase\b', content))
        complexity += len(re.findall(r'\bcatch\b', content))
        complexity += len(re.findall(r'\b&&\b', content))
        complexity += len(re.findall(r'\b\|\|\b', content))

        return complexity

    def _calculate_priority(self, issues: List[Dict]) -> str:
        """Calculate priority based on issue severity"""
        severity_counts = {
            'critical': 0,
            'high': 0,
            'medium': 0,
            'low': 0
        }

        for issue in issues:
            severity = issue.get('severity', 'low')
            severity_counts[severity] = severity_counts.get(severity, 0) + 1

        if severity_counts['critical'] > 0:
            return 'critical'
        elif severity_counts['high'] > 2:
            return 'high'
        elif severity_counts['medium'] > 5:
            return 'medium'
        else:
            return 'low'
